src/pyodide/patch_pyodide_js.bzl

```diff
@@ -0,0 +1,74 @@
+"""
+This file patches various details in pyodide.asm.js. We should eventually switch
+to using `.patch` files here and patch the sources of these upstream files. Some
+of these patches can be resolved by passing an extra flag to the Emscripten
+linker. `Date.now()` and `WebAssembly.Module()` are from Emscripten's runtime
+libraries injected at link time, so we can also patch them in Emscripten.
+
+Because the file we are patching is a minified generated file, it's not
+appropriate for patch files since the patch context is likely to change due to
+stuff that doesn't matter for our sake. In fact, if we just want to substitute
+`Date.now()` for `monotonicDateNow()` patch context would be extra stuff we'd
+have to maintain for not much benefit.
+
+1. `pyodide.asm.js` exposes `_createPyodideModule` via assignment to
+   `globalThis` because module workers in Firefox have < 1 year of support. This
+   changes it to an es6 export. In the future this would better be accomplished
+   by passing `-sEXPORT_ES6` to Emscripten's linker.
+
+2. Pyodide needs to synchronously compile webassembly code with `new
+   WebAssembly.Module()`. We replace these with calls to `newWasmModule` from
+   `internal:unsafe-eval`. To be paranoid, `builtin_wrappers` adds an extra
+   check to `newWasmModule` which checks that the call comes from
+   `convertJsFunctionToWasm`. See comment about this in
+   `internal/builtin_wrappers.js`.
+
+3. Python uses the current time as a cache key to check whether a directory is
+   modified. It needs modifying a directory to change the modification time of
+   the directory. The disabled timers interfere with this. `monotonicDateNow` is
+   a wrapper around `Date.now()` that always increments the returned time by at
+   least one millisecond.
+
+4. Various junk that is linked into Pyodide uses `addEventListener` in a way
+   that throws errors so we dummy it out. This junk is unnecessary for us and
+   wasting large amounts of initialization time, when we set up our own Pyodide
+   build we'll get rid of it and this won't be necessary.
+"""
+
+PRELUDE = """
+import { newWasmModule, monotonicDateNow } from "pyodide-internal:builtin_wrappers";
+
+function addEventListener(){}
+"""
+
+REPLACEMENTS = [
+    ["var _createPyodideModule", PRELUDE + "export const _createPyodideModule"],
+    ["globalThis._createPyodideModule = _createPyodideModule;", ""],
+    ["new WebAssembly.Module", "newWasmModule"],
+    ["Date.now", "monotonicDateNow"],
+]
+
+def _impl(ctx):
+    args = ctx.actions.args()
+    args.add_all(ctx.files.input)
+    args.add(ctx.outputs.output)
+
+    ctx.actions.expand_template(
```

Comment:
this could be a simple macro, the rule is not necessary. sorry :)

Reply:
I tried using `expand_template` directly and I got a name error. Do macros get access to `ctx.actions`? Or how do they get to `expand_template`?

Reply:
you just 

`load("@bazel_skylib//rules:expand_template.bzl", "expand_template")`

and use `expand_template()` directly as a rule in your macro. Or are you talking about something else?

Reply:
It seems that the expand template rule was first added in `bazel_skylib` v1.3.0:
https://github.com/bazelbuild/bazel-skylib/commit/2a87d4a62af886fb320883aba102255aba87275e

We're using `bazel_skylib` v1.2.1 currently. But I guess it should be easy enough to update.